### PR TITLE
Add webRTC proxy

### PR DESCRIPTION
**Pull request by isherman**
A quick experiment in proxying UDP and RTP traffic to/from webRTC.

Instructions in [go/webrtc/README.md](go/webrtc/README.md)

I'm not sure about the choice of golang. I'm no expert in it, and the more languages added here, the more overhead there is to tooling, code review, etc. But I know [pion](https://github.com/pion/webrtc) is a high-quality webRTC library, and go feels like a good choice for this sort of component.

This approach should work for peers on the same LAN. The web browser acts as the offerer, and sends its local ICE candidates (in this case, basically its IP address on the LAN) in the initial SDP offer. The proxy returns its local ICE candidates  (its IP address) in its answer. So a single synchronous RPC is enough to exchange the necessary "signals". 

In a WAN scenario, the browser and proxy need to be able to send (aka "trickle") ICE candidates to each other as they are discovered, asynchronously. Which is handled by an out-of-band signaling channel (something in the cloud, or a carrier pigeon). But other than that the proxy code would remain largely the same.

### CODE REVIEW

File: scripts/build-protos.sh [hunk 1]
```diff
@@ -1,10 +1,12 @@
-#!/bin/bash
+#!/usr/local/bin/bash
```
Comment (ethanrublee) [inline scripts/build-protos.sh:1]: this change to the shebang work on my machine, and make doesn't give a useful error message:

```
scripts/build-protos.sh
make: scripts/build-protos.sh: Command not found
Makefile:7: recipe for target 'protos' failed
make: *** [protos] Error 127
```

File: go/webrtc/README.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+```bash
+# install go
+# (tested with 1.14, where go modules are enabled by default)
+
+# build protobufs and twirp server stubs for Go
+make protos
+
+# start the twirp server
+cd go/webrtc
+go run cmd/proxy-server/main.go
```
Comment (ethanrublee) [inline go/webrtc/README.md:12]: When running this command i get some path errors:
```
$ go run cmd/proxy-server/main.go
warning: GOPATH set to GOROOT (/home/rublee/code/ian-tractor/env/go) has no effect
cmd/proxy-server/main.go:8:2: cannot find package "github.com/farm-ng/tractor/genproto" in any of:
	/home/rublee/code/ian-tractor/env/go/src/github.com/farm-ng/tractor/genproto (from $GOROOT)
	($GOPATH not set. For more details see: 'go help gopath')
cmd/proxy-server/main.go:9:2: cannot find package "github.com/farm-ng/tractor/webrtc/internal/server" in any of:
	/home/rublee/code/ian-tractor/env/go/src/github.com/farm-ng/tractor/webrtc/internal/server (from $GOROOT)
	($GOPATH not set. For more details see: 'go help gopath')
cmd/proxy-server/main.go:6:2: cannot find package "github.com/rs/cors" in any of:
	/home/rublee/code/ian-tractor/env/go/src/github.com/rs/cors (from $GOROOT)
	($GOPATH not set. For more details see: 'go help gopath')
```

```
(env) rublee@farmws01:~/code/ian-tractor/go/webrtc$ which go
/home/rublee/code/ian-tractor/env/go/bin/go
(env) rublee@farmws01:~/code/ian-tractor/go/webrtc$ go version
warning: GOPATH set to GOROOT (/home/rublee/code/ian-tractor/env/go) has no effect
go version go1.15 linux/amd64
```

File: scripts/build-protos.sh (hunk 1 above)
Comment (isherman) [reply scripts/build-protos.sh:1]: ack, sorry about that. I made the change to support that `shopt -s globstar` on my machine but I can find a better fix.

File: go/webrtc/README.md (hunk 2 above)
Comment (ethanrublee) [reply go/webrtc/README.md:12]: I've also tried with 1.14, but get similar errors:

```
(env) rublee@farmws01:~/code/ian-tractor/go/webrtc$ go run cmd/proxy-server/main.go
cmd/proxy-server/main.go:8:2: cannot find package "github.com/farm-ng/tractor/genproto" in any of:
	/home/rublee/code/ian-tractor/env/go/src/github.com/farm-ng/tractor/genproto (from $GOROOT)
	/home/rublee/code/ian-tractor/go/src/github.com/farm-ng/tractor/genproto (from $GOPATH)
cmd/proxy-server/main.go:9:2: cannot find package "github.com/farm-ng/tractor/webrtc/internal/server" in any of:
	/home/rublee/code/ian-tractor/env/go/src/github.com/farm-ng/tractor/webrtc/internal/server (from $GOROOT)
	/home/rublee/code/ian-tractor/go/src/github.com/farm-ng/tractor/webrtc/internal/server (from $GOPATH)
cmd/proxy-server/main.go:6:2: cannot find package "github.com/rs/cors" in any of:
	/home/rublee/code/ian-tractor/env/go/src/github.com/rs/cors (from $GOROOT)
	/home/rublee/code/ian-tractor/go/src/github.com/rs/cors (from $GOPATH)
(env) rublee@farmws01:~/code/ian-tractor/go/webrtc$ go version
go version go1.14 linux/amd64
(env) rublee@farmws01:~/code/ian-tractor/go/webrtc$ which go
/home/rublee/code/ian-tractor/env/go/bin/go
(env) rublee@farmws01:~/code/ian-tractor/go/webrtc$ echo $GOPATH
/home/rublee/code/ian-tractor/go

```

File: go/webrtc/README.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+# Send some test video
+gst-launch-1.0 videotestsrc ! 'video/x-raw, width=640, height=480' ! videoconvert ! video/x-raw,format=I420 ! vp8enc error-resilient=partitions keyframe-max-dist=10 auto-alt-ref=true cpu-used=5 deadline=1 ! rtpvp8pay ! udpsink host=127.0.0.1 port=5004
+# OR
+ffmpeg -re -f lavfi -i testsrc=size=640x480:rate=30 -vcodec libvpx -cpu-used 5 -deadline 1 -g 10 -error-resilient 1 -auto-alt-ref 1 -f rtp rtp://127.0.0.1:5004
+
+# Run https://jsfiddle.net/y5aq6hnm/1/
```
Comment (ethanrublee) [inline go/webrtc/README.md:19]: i get this error after some time, not sure why yet, pasting here for feedback:

```
Message from DataChannel: This is my DataChannel message!
Connection State has changed failed 
panic: failed to send packet: write udp 192.168.1.158:51068->192.168.1.158:55862: use of closed network connection

goroutine 148 [running]:
github.com/farm-ng/tractor/webrtc/internal/proxy.(*Proxy).Start.func3(0xc000010598, 0xc0001f9000, 0x1000, 0x1000, 0x60, 0xc000117720)
	/home/rublee/code/ian-tractor/go/webrtc/internal/proxy/main.go:173 +0x1cc
created by github.com/farm-ng/tractor/webrtc/internal/proxy.(*Proxy).Start
	/home/rublee/code/ian-tractor/go/webrtc/internal/proxy/main.go:158 +0x830
exit status 2
```

File: go/webrtc/README.md (hunk 2 above)
Comment (ethanrublee) [reply go/webrtc/README.md:12]: the module fix worked, thanks!

**Comment by isherman** [comment]
@ethanrublee also suggests testing `h264` with `gst-launch-1.0 v4l2src ! videoconvert ! x264enc ! video/x-h264, stream-format=byte-stream ! rtph264pay pt=96 mtu=1400 config-interval=10 ! udpsink host=127.0.0.1  port=5004`

**Comment by ethanrublee** [comment]
I was able to do vp8 encoding on the nano accelerated using this pipeline:
```
gst-launch-1.0 v4l2src ! videoconvert ! omxvp8enc bitrate=1000000 !  rtpvp8pay ! udpsink host=0.0.0.0  port=5004
```

h264 pipeline (hasn't worked yet, but for reference)
```
gst-launch-1.0 v4l2src ! videoconvert ! omxh264enc control-rate=1 bitrate=2500000 ! video/x-h264, stream-format=byte-stream ! rtph264pay pt=96 mtu=1400 config-interval=10 ! udpsink host=0.0.0.0  port=5004
```

**Comment by ethanrublee** [comment]
For reference, i've added the gst-webrtc example that's verified working on the nano (https://github.com/farm-ng/tractor/pull/31/commits/a60352c30cb503e2a78eefe10165d3810d5d8ebb), under apps/gst-webrtc.  We can throw it out, but i've been able to get that to work consistently and it has low CPU load.

I like the idea of being able to do the datachannels and the video streams from one location, as you have it here, so consider it just design reference for now.

File: go/webrtc/README.md (hunk 3 above)
Comment (isherman) [reply go/webrtc/README.md:19]: Addressed by https://github.com/farm-ng/tractor/pull/33/files#diff-0bd6c1259358eec9058c394d03552e0cR156

**Comment by isherman** [comment]
Would still benefit from some refactoring, especially on the go side, but I think it's in a decent place -- will keep iterating.

Since the original PR description, has expanded to include:
- A go-ipc implementation 
- Frontend visualization of apriltag annotations

File: app/frontend/src/index.tsx [hunk 4]
```diff
@@ -70,7 +71,11 @@ export class App extends React.Component<{}, State> {
 
   public render(): any {
     return (
-      <Scene />
+      <React.Fragment>
+        <Rtc />
+        <Scene />
```
Comment (ethanrublee) [inline app/frontend/src/index.tsx:76]: Any way to exit the scene to use the webrtc datachannel rather than websocket?

**Review by ethanrublee — APPROVED** [review]
Ran through this on the tractor, seems to work well for a first pass.

A similar install script for the workstation similar to what you have from jetson would be appreciated.

Good to merge in my opinion.

File: app/frontend/src/index.tsx (hunk 4 above)
Comment (isherman) [reply app/frontend/src/index.tsx:76]: For sure. Let me do that in a follow-up change -- tracking in https://github.com/farm-ng/tractor/issues/63#issuecomment-687455735

**Pull request merged**